### PR TITLE
feat(freetype): merge freetype outline and image into one context

### PR DESCRIPTION
**Pull request by W-Mai**
### Description of the feature or fix

Merge freetype outline and image into one context.

Now you can create bitmap based freetype font and vector graph based outline freetype font at same time by using these flag:

```cpp
enum {
    LV_FREETYPE_FONT_RENDER_MODE_BITMAP = 0,
    LV_FREETYPE_FONT_RENDER_MODE_OUTLINE = 1,
};
```cpp
// create a bitmap based freetype font
lv_freetype_font_create("./res/MiSans-Regular.ttf", LV_FREETYPE_FONT_RENDER_MODE_BITMAP,
                                                24,
                                                LV_FREETYPE_FONT_STYLE_ITALIC);

// create a vector graph freetype font
lv_freetype_font_create("./res/MiSans-Regular.ttf", LV_FREETYPE_FONT_RENDER_MODE_OUTLINE,
                                                24,
                                                LV_FREETYPE_FONT_STYLE_NORMAL);
```

snapshot:

image

## Note:
If your environment doesn't support vector drawing, you will get a result like this
image

### Checkpoints
- [x] Run `code-format.py` from the scripts folder. [astyle](http://astyle.sourceforge.net/install.html) needs to be installed.
- [ ] Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed
- [ ] Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant. 
- [ ] Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- [ ] If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/release/v8.3/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/release/v8.3/Kconfig).
 
Be sure the following conventions are followed:
- [ ] Follow the [Styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Prefer `enum`s instead of macros. If inevitable to use `define`s export them with `LV_EXPORT_CONST_INT(defined_value)` right after the `define`.
- [ ] In function arguments prefer `type name[]` declaration for array parameters instead of `type * name`
- [ ] Use typed pointers instead of `void *` pointers
- [ ] Do not `malloc` into a static or global variables. Instead declare the variable in `lv_global_t` structure in [`lv_global.h`](https://github.com/lvgl/lvgl/blob/master/src/core/lv_global.h) and mark the variable with `(LV_GLOBAL_DEFAULT()->variable)` when it's used. See a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#memory-management).
- [ ] Widget constructor must follow the `lv_<widget_name>_create(lv_obj_t * parent)` pattern.
- [ ] Widget members function must start with `lv_<module_name>` and should receive `lv_obj_t *` as first argument which is a pointer to widget object itself.  
- [ ] `struct`s should be used via an API and not modified directly via their elements.
- [ ] `struct` APIs should follow the widgets' conventions. That is to receive a pointer to the `struct` as the first argument, and the prefix of the `struct` name should be used as the prefix of the function name too (e.g.  `lv_disp_set_default(lv_disp_t * disp)`)
- [ ] Functions and `struct`s which are not part of the public API must begin with underscore in order to mark them as "private".
- [ ] Arguments must be named in H files too.
- [ ] To register and use callbacks one of the following needs to be followed (see a detailed description [here](https://docs.lvgl.io/master/get-started/bindings/micropython.html#callbacks)):
  - For both the registration function and the callback pass a pointer to a `struct` as the first argument. The `struct` must contain `void * user_data` field.
  - The last argument of the registration function must be `void * user_data` and the same `user_data` needs to be passed as the last argument of the callback.
  - Callback types not following these conventions should end with `xcb_t`.

### CODE REVIEW

**Comment by W-Mai** [comment]
I find that there is always a problem with my pr. The pr page on GitHub after force push does not respond, so I need to reset codebase manually, and then codebase will not go to rebase based on the latest master, but the commit I submitted at that time, which is weird.

This pr is create a fork one from https://github.com/lvgl/lvgl/pull/5146

**Comment by kisvegabor** [comment]
I couldn't test it with VG-Lite, so I have only some high level comments:
- Can we use vector fonts with ThorVG too? If so could you add support for it as well?
- Now that [freetype is installed](https://github.com/lvgl/lvgl/pull/5155) please add some tests too.

**Comment by W-Mai** [comment]
@FASTSHIFT has implemented a porting layer from ThorVG to VGLite. I test this feature on simulator with his codes.

**Comment by FASTSHIFT** [comment]
> I couldn't test it with VG-Lite, so I have only some high level comments:
> 
> * Can we use vector fonts with ThorVG too? If so could you add support for it as well?
> * Now that [freetype is installed](https://github.com/lvgl/lvgl/pull/5155) please add some tests too.

I implemented a simulated software VG-Lite implementation: https://github.com/FASTSHIFT/vg_lite_tvg.
Using this tool can greatly speed up GPU debugging. We can consider adding it to CI, so that modifications to the framework can be quickly verified and problems fixed, reducing the time spent debugging on the hardware.

File: src/draw/lv_draw_label.h [hunk 1]
```diff
@@ -77,6 +77,7 @@ typedef struct {
 typedef enum {
     LV_DRAW_LETTER_BITMAP_FORMAT_A8,
     LV_DRAW_LETTER_BITMAP_FORMAT_IMAGE,
+    LV_DRAW_LETTER_BITMAP_FORMAT_VECTOR,
```
Comment (XuNeo) [inline src/draw/lv_draw_label.h]: BITMAP VECTOR is kind weird.

File: src/libs/freetype/lv_freetype.c [hunk 2]
```diff
@@ -96,16 +101,15 @@ lv_result_t lv_freetype_init(uint32_t max_faces, uint32_t max_sizes, uint32_t ma
...
-
-    if(!ctx->cache_context) {
-        LV_LOG_ERROR("cache context create failed");
-        lv_freetype_cleanup(ctx);
-        return LV_RESULT_INVALID;
-    }
-
     _lv_ll_init(&ctx->face_id_ll, sizeof(face_id_node_t));
 
+    lv_cache_ops_t ops = {
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype.c:106]:
```suggestion
    static const lv_cache_ops_t ops = {
```

File: src/libs/freetype/lv_freetype.c [hunk 3]
```diff
@@ -313,4 +352,67 @@ static void lv_freetype_drop_face_id(lv_freetype_context_t * ctx, FTC_FaceID fac
     LV_ASSERT_MSG(false, "face_id not found");
 }
 
+/*-----------------
+ * Cache Node Cache Callbacks
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype.c:356]: Could you comment on what's a cache node cache?

File: src/libs/freetype/lv_freetype_glyph.c [hunk 4]
```diff
@@ -46,10 +45,8 @@ static lv_cache_compare_res_t freetype_glyph_compare_cb(const lv_freetype_glyph_
 /**********************
  *   GLOBAL FUNCTIONS
  **********************/
-lv_cache_t * lv_freetype_glyph_cache_create(lv_freetype_font_dsc_t * dsc)
+lv_cache_t * lv_freetype_glyph_cache_create(void)
 {
-    LV_ASSERT_FREETYPE_FONT_DSC(dsc);
-
     lv_cache_ops_t ops = {
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype_glyph.c:50]:
```suggestion
    static const lv_cache_ops_t ops = {
```

File: src/libs/freetype/lv_freetype_image.c [hunk 5]
```diff
@@ -176,7 +127,7 @@ static void freetype_image_release_cb(const lv_font_t * font, lv_font_glyph_dsc_
 {
     LV_ASSERT_NULL(font);
     lv_freetype_font_dsc_t * dsc = (lv_freetype_font_dsc_t *)font->dsc;
-    lv_cache_release(dsc->cache_node->image_cache, g_dsc->entry, NULL);
+    lv_cache_release(dsc->cache_node->draw_data_cache, g_dsc->entry, NULL);
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype_image.c:123]: I personally find it hard to understand these words `dsc->cache_node->draw_data_cache`. I would like to hear from others opinions.

File: src/libs/freetype/lv_freetype_outline.c [hunk 6]
```diff
@@ -87,84 +60,38 @@ static lv_cache_compare_res_t freetype_glyph_outline_cmp_cb(const lv_freetype_ou
...
-
-    /* Must ensure that cache_node is not used */
-    LV_ASSERT(_lv_ll_get_len(cache_ll) == 0);
-
-    LV_LOG_INFO("cache_context = %p", cache_context);
-    lv_free(cache_context);
-}
-
-bool lv_freetype_on_font_create(lv_freetype_font_dsc_t * dsc)
+bool lv_freetype_outline_font_create(lv_freetype_font_dsc_t * dsc)
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype_outline.c:63]: Add a comment above to explain what does return value `false` mean.

File: src/libs/freetype/lv_freetype_outline.c [hunk 7]
```diff
@@ -394,18 +213,19 @@ static void ft_vector_to_lv_vector(lv_freetype_outline_vector_t * dest, const FT
...
+        LV_LOG_ERROR("event_cb is not set");
         return LV_RESULT_INVALID;
     }
 
     lv_event_t e;
     lv_memzero(&e, sizeof(e));
     e.code = code;
     e.param = param;
-    e.user_data = ctx->user_data;
+    e.user_data = NULL;
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype_outline.c:232]: Remove?

File: src/libs/freetype/lv_freetype_private.h [hunk 8]
```diff
@@ -60,24 +60,43 @@ extern "C" {
  *      TYPEDEFS
  **********************/
 
-typedef struct _lv_freetype_cache_context_t lv_freetype_cache_context_t;
 typedef struct _lv_freetype_cache_node_t lv_freetype_cache_node_t;
 
+struct _lv_freetype_cache_node_t {
+    const char * pathname;
```
Comment (XuNeo) [inline src/libs/freetype/lv_freetype_private.h:66]: If I understand correctly, the cache node stores glyph data. Do we need to store pathname here?

Comment (W-Mai) [reply src/libs/freetype/lv_freetype_private.h:66]: for cache searching

Comment (W-Mai) [reply src/libs/freetype/lv_freetype_private.h:66]: not only for glyph data, but also glyph and draw data cache

File: src/draw/lv_draw_label.h (hunk 1 above)
Comment (W-Mai) [reply src/draw/lv_draw_label.h]: sure

File: src/libs/freetype/lv_freetype.c (hunk 3 above)
Comment (W-Mai) [reply src/libs/freetype/lv_freetype.c:356]: @FASTSHIFT named this name before

File: src/draw/lv_draw_label.h (hunk 1 above)
Comment (XuNeo) [reply src/draw/lv_draw_label.h]: I have changed it to `LV_DRAW_LETTER_GLYPH_FORMAT_A8 ` in 
https://github.com/lvgl/lvgl/pull/5056/files#diff-ebd8930142aae725f59a3fc797b3fa540cdca53ec7fe17e614d1056ed8daf84e 
Please check if it's suitable.

**Comment by W-Mai** [comment]
testcase here https://github.com/lvgl/lvgl/pull/5196

File: src/draw/lv_draw_label.h [hunk 9]
```diff
@@ -77,6 +77,7 @@ typedef struct {
 typedef enum {
     LV_DRAW_LETTER_BITMAP_FORMAT_A8,
     LV_DRAW_LETTER_BITMAP_FORMAT_IMAGE,
+    LV_DRAW_LETTER_VECTOR_FORMAT,
 } lv_draw_glyph_bitmap_format_t;
```
Comment (kisvegabor) [inline src/draw/lv_draw_label.h:81]: This type should be removed and `lv_font_glyph_format_t` should be used instead. (See below).

File: src/font/lv_font.h [hunk 10]
```diff
@@ -28,6 +28,7 @@ extern "C" {
 
 /* imgfont identifier */
 #define LV_IMGFONT_BPP 9
+#define LV_VECFONT_BPP 0
```
Comment (kisvegabor) [inline src/font/lv_font.h:31]: Using the BPP for marking format is hacky. To make it more generic we should rename `font->bpp` to `font->glyph_format`. 
Let's make it an `enum` like:
```c
typedef enum {
  LV_FONT_GYLPH_FORMAT_A1 = 1,
  LV_FONT_GYLPH_FORMAT_A2 = 2,
  LV_FONT_GYLPH_FORMAT_A4 = 4,
  LV_FONT_GYLPH_FORMAT_A8 = 8,
  LV_FONT_GYLPH_FORMAT_VECTOR = 9,
  LV_FONT_GYLPH_FORMAT_IMAGE = 10,
}lv_font_glyph_format_t;
```

It required to update font converter but we are still in time to do that.

Comment (W-Mai) [reply src/font/lv_font.h:31]: OK, it's great!

I think we can create a new PR to handle this font format things to make it more clear.

File: src/libs/freetype/lv_freetype_image.c (hunk 5 above)
Comment (W-Mai) [reply src/libs/freetype/lv_freetype_image.c:123]: Cache of data to be drawn

**Comment by W-Mai** [comment]
It seems that everything is done, can we merge this pr first? The next step is to continue to improve the code.😁

**Review by FASTSHIFT — DISMISSED** [review]

**Review by FASTSHIFT — DISMISSED** [review]

**Review by FASTSHIFT — DISMISSED** [review]

**Review by dqdroj9303 — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Pull request merged**